### PR TITLE
Support for NCLOB datatype

### PR DESCRIPTION
**Pull request by luke2237**
Added support for NCLOB data types
Updated JDBC connection to handle nclob fields
Added nclob->ntext type to native_database_types.
Updated Specs for NCLOB. They are green.

### CODE REVIEW

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb [hunk 1]
```diff
@@ -882,6 +882,234 @@ class ::TestSerializeEmployee < ActiveRecord::Base
...
+      create_table :test_serialize_employees, force: true do |t|
+        t.string  :first_name, limit: 20
+        t.string  :last_name, limit: 25
+      end
+      add_column :test_serialize_employees, :comments, :ntext
+    end
+
+    @nchar_data = "日オメヌ"
+    @nchar_data2 = "使セ転"
+
```
Comment (yahonda) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: What do these two `nchar_data` and `nchar_data2` represent?  These values look meaningless and garbled for me. Note : I speak Japanese and English.

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb [hunk 2]
```diff
@@ -882,6 +882,234 @@ class ::TestSerializeEmployee < ActiveRecord::Base
...
+    @employee = TestEmployee.create!(
+      comments: :test_comment
+    )
+    expect(@employee).to be_valid
+  end
+
+  it "should respect attr_readonly setting for NCLOB column" do
+    @employee = TestEmployeeReadOnlyNClob.create!(
+      first_name: "First",
+      comments: "日オメヌロ"
```
Comment (yahonda) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: Another meaningless and garbled character (for me).

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb [hunk 3]
```diff
@@ -882,6 +882,234 @@ class ::TestSerializeEmployee < ActiveRecord::Base
...
+    )
+    expect(@employee).to be_valid
+    @employee.reload
+    expect(@employee.comments).to eq("日オメヌロ")
+    @employee.comments = "勢はう"
```
Comment (yahonda) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: Another meaningless and garbled character (for me).

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb [hunk 4]
```diff
@@ -882,6 +882,234 @@ class ::TestSerializeEmployee < ActiveRecord::Base
...
+    @employee.save
+    @employee.reload
+    expect(@employee.comments).to eq(ruby_data2)
+  end
+
+  it "should keep unchanged serialized data when other columns changed" do
+    @employee = Test2Employee.create!(
+      first_name: "First",
+      last_name: "Last",
+      comments: "福れは"
```
Comment (yahonda) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: Another meaningless and garbled character (for me).

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb [hunk 5]
```diff
@@ -882,6 +882,234 @@ class ::TestSerializeEmployee < ActiveRecord::Base
...
+  include SchemaSpecHelper
+
+  before(:all) do
+    ActiveRecord::Base.establish_connection(CONNECTION_PARAMS)
+    @conn = ActiveRecord::Base.connection
+    schema_define do
+      create_table :test_employees, force: true do |t|
+        t.string  :first_name, limit: 20
+        t.string  :last_name, limit: 25
+        t.column  :comments, :ntext
```
Comment (yahonda) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: I think `t.ntext   :comments` looks better for migrations.  you can refer the commit below to support `t.ntext` for migrations.

https://github.com/yahonda/oracle-enhanced/commit/7ee3bb367d3ce450adb8ae136ee59b0624cac40d

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb [hunk 6]
```diff
@@ -882,6 +882,234 @@ class ::TestSerializeEmployee < ActiveRecord::Base
...
+      end
+      create_table :test2_employees, force: true do |t|
+        t.string  :first_name, limit: 20
+        t.string  :last_name, limit: 25
+        t.column  :comments, :ntext
```
Comment (yahonda) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: Better to use `t.ntext` here also.

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb (hunk 1 above)
Comment (luke2237) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: These are random multibyte characters used to be stored in nclob datatype. I picked them off internet. Think of them as Lorem Ipsum.

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb (hunk 2 above)
Comment (luke2237) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: Once again some random Japanese multibyte characters.

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb (hunk 3 above)
Comment (luke2237) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: Same as above.

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb (hunk 4 above)
Comment (luke2237) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: Same as above

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb (hunk 5 above)
Comment (luke2237) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: It does look better. I was getting error and had not updated schema_definitions. Thank you for your help!

File: spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb (hunk 6 above)
Comment (luke2237) [inline spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb]: I agree.

**Comment by luke2237** [comment]
I've updated the code based on the comments. Please review.

**Comment by yahonda** [comment]
Thanks for the update. Please add N-quote literal support and specs.

*  National character set requires N-quote. 

http://docs.oracle.com/database/122/SQLRF/Literals.htm#GUID-1824CBAA-6E16-4921-B2A6-112FB02248DA

*** Oracle enhanced adapter `NCHAR` sql type implements like this.

https://github.com/rsim/oracle-enhanced/blob/master/lib/active_record/connection_adapters/oracle_enhanced/quoting.rb#L96-L97
```ruby
          when ActiveRecord::OracleEnhanced::Type::NationalCharacterString::Data then
            "N".dup << "'#{quote_string(value.to_s)}'"
````

*** Specs for N quoted literals

https://github.com/rsim/oracle-enhanced/blob/master/spec/active_record/connection_adapters/oracle_enhanced_data_types_spec.rb#L1155-L1164
```ruby

  it "should quote with N prefix" do
    columns = @conn.columns("test_items")
    %w(nchar_column nvarchar2_column char_column varchar2_column).each do |col|
      column = columns.detect { |c| c.name == col }
      value = @conn.type_cast_from_column(column, "abc")
      expect(@conn.quote(value)).to eq(column.sql_type[0, 1] == "N" ? "N'abc'" : "'abc'")
      nilvalue = @conn.type_cast_from_column(column, nil)
      expect(@conn.quote(nilvalue)).to eq("NULL")
    end
  end
```


* Add specs `db:schema:dump` to generate `ntext` type correctly
* Add specs `db:structure:dump` to generate `NCLOB` sql type correctly

Thanks.

**Comment by luke2237** [comment]
Made changes as recommended in the review. The Quoting for NCLOB can not be the same as NCHAR as it is a large object. Added appropriate quoting and specs. Please review and let me know in case something needs to be changed. 

Thank you!

File: spec/active_record/connection_adapters/oracle_enhanced_structure_dump_spec.rb [hunk 7]
```diff
@@ -152,6 +152,19 @@ class ::TestPost < ActiveRecord::Base
       expect(dump).to match(/CREATE TABLE \"BARS\" \(\n \"ID\" NUMBER\(38,0\) NOT NULL,\n \"SUPER\" RAW\(255\) GENERATED ALWAYS AS \(HEXTORAW\(TO_CHAR\(ID\)\)\) VIRTUAL/)
     end
 
+    it "should dump NCLOB columns" do
+      skip "Not supported in this database version" unless @oracle11g_or_higher
```
Comment (yahonda) [inline spec/active_record/connection_adapters/oracle_enhanced_structure_dump_spec.rb]: NCLOB data type exists at least Oracle 9i 9.0.1. Please remove this `skip` statement.

**Comment by luke2237** [comment]
Done. Made the changes and squashed into 1 commit.

**Pull request closed without merging**